### PR TITLE
[vision] Acquired initial game data and converted it into sendable data to ros

### PR DESCRIPTION
**Pull request by Unagifan**
Acquired initial game data and converted it into sendable data to ros. Not sure how John wanted to receive the scoring placement data for his side, so I kept everything within the offboard subsystem

### CODE REVIEW

**Review by JohnGilb — COMMENTED** [review]
@WasabiFan should probably take a look at this - I haven't been involved in all this bit-level hacking =]

File: Competition2018/src/main/java/competition/subsystems/offboard/OffboardFramePackingUtils.java [hunk 1]
```diff
@@ -21,7 +21,22 @@
...
     }
-}
+    
+    public static byte[] packScoringPlacement(int sequence) {
+        return new byte[] {
+                (byte)(sequence & 0xFF)    
+        };
+    }
+    
+    public static byte[] packHeadingFrame(double headingDegrees) {
```
Comment (WasabiFan) [inline Competition2018/src/main/java/competition/subsystems/offboard/OffboardFramePackingUtils.java]: Where'd this come from?

File: Competition2018/src/main/java/competition/subsystems/offboard/OffboardInterfaceSubsystem.java [hunk 2]
```diff
@@ -83,13 +87,37 @@ public void sendSetCurrentCommand(int commandId) {
...
+            else if(MatchData.getOwnedSide(feature).equals(MatchData.OwnedSide.RIGHT)) {
+                sequence.add(1);
+            }
+            else {
+                //Insert Error Message as it is unknown
+            }
+        }
+        int total = 0;
+        for (Integer i : sequence) { 
+            total = 10*total + i;
```
Comment (WasabiFan) [inline Competition2018/src/main/java/competition/subsystems/offboard/OffboardInterfaceSubsystem.java]: I would suggest a slightly different approach. It's the same concept, but works a bit better with the binary internal representation. You can use a "bit shift" `>>` to store each value as a single binary bit. Additionally, I'd suggest moving this logic into the "pack" method. That method should take the side info and figure out how to do the bit manipulation.

File: Competition2018/src/main/java/competition/subsystems/offboard/OffboardFramePackingUtils.java (hunk 1 above)
Comment (Unagifan) [reply Competition2018/src/main/java/competition/subsystems/offboard/OffboardFramePackingUtils.java]: No clue? Should I delete it?

File: Competition2018/src/main/java/competition/subsystems/offboard/OffboardInterfaceSubsystem.java (hunk 2 above)
Comment (Unagifan) [reply Competition2018/src/main/java/competition/subsystems/offboard/OffboardInterfaceSubsystem.java]: Thanks!

**Review by JohnGilb — APPROVED** [review]

File: Competition2018/src/main/java/competition/subsystems/offboard/OffboardFramePackingUtils.java [hunk 3]
```diff
@@ -20,7 +22,19 @@
     
     public static byte[] packSetCommandFrame(int commandId) {
         return new byte[] {
-            (byte)(commandId & 0xFF)
+                (byte)(commandId & 0xFF)
```
Comment (WasabiFan) [inline Competition2018/src/main/java/competition/subsystems/offboard/OffboardFramePackingUtils.java]: Undo?

File: Competition2018/src/main/java/competition/subsystems/offboard/OffboardFramePackingUtils.java [hunk 4]
```diff
@@ -20,7 +22,19 @@
...
         };
     }
-}
+    
+    public static byte[] packScoringPlacement(ArrayList<Integer> scoringPlacement) {
+    	int sequence = 0;
+    	for(int i = 0; i < 3; i++) {
+    		int scoring = (scoringPlacement.get(i) << (2-i));
```
Comment (WasabiFan) [inline Competition2018/src/main/java/competition/subsystems/offboard/OffboardFramePackingUtils.java]: The `2-` part is a bit weird; I'd suggest changing the loop to be `int i = 3; i >= 1; i--`

File: Competition2018/src/main/java/competition/subsystems/offboard/OffboardInterfaceSubsystem.java [hunk 5]
```diff
@@ -112,6 +116,25 @@ public void sendSetCurrentCommand(int commandId) {
         return queueContents;
     }
     
+    private void sendScoringPlacement() {
+        //Sequence - Near, Scale, Far
+        //Left - 1, Right - 0
+        ArrayList<Integer> sequence = new ArrayList<Integer>();
+        for(MatchData.GameFeature feature: MatchData.GameFeature.values()) {
```
Comment (WasabiFan) [inline Competition2018/src/main/java/competition/subsystems/offboard/OffboardInterfaceSubsystem.java]: Can you move this logic to the `pack...` method? This is internal logic which I think fits better over there.

File: Competition2018/src/main/java/competition/subsystems/offboard/OffboardInterfaceSubsystem.java [hunk 6]
```diff
@@ -112,6 +116,25 @@ public void sendSetCurrentCommand(int commandId) {
...
+            if(MatchData.getOwnedSide(feature).equals(MatchData.OwnedSide.LEFT)) {
+                sequence.add(1);
+            }
+            else if(MatchData.getOwnedSide(feature).equals(MatchData.OwnedSide.RIGHT)) {
+                sequence.add(0);
+            }
+            else {
+                //Insert Error Message as it is unknown
```
Comment (WasabiFan) [inline Competition2018/src/main/java/competition/subsystems/offboard/OffboardInterfaceSubsystem.java]: Log

**Comment by WasabiFan** [comment]
@Unagifan You'll need to pull before continuing work on this branch

File: Competition2018/src/main/java/competition/subsystems/offboard/OffboardFramePackingUtils.java (hunk 3 above)
Comment (Unagifan) [reply Competition2018/src/main/java/competition/subsystems/offboard/OffboardFramePackingUtils.java]: !

File: Competition2018/src/main/java/competition/subsystems/offboard/OffboardInterfaceSubsystem.java (hunk 5 above)
Comment (Unagifan) [reply Competition2018/src/main/java/competition/subsystems/offboard/OffboardInterfaceSubsystem.java]: !

File: Competition2018/src/main/java/competition/subsystems/offboard/OffboardInterfaceSubsystem.java (hunk 6 above)
Comment (Unagifan) [reply Competition2018/src/main/java/competition/subsystems/offboard/OffboardInterfaceSubsystem.java]: !

File: Competition2018/src/main/java/competition/subsystems/offboard/OffboardFramePackingUtils.java (hunk 4 above)
Comment (Unagifan) [reply Competition2018/src/main/java/competition/subsystems/offboard/OffboardFramePackingUtils.java]: !

**Comment by Unagifan** [comment]
@WasabiFan

File: Competition2018/src/main/java/competition/subsystems/offboard/OffboardInterfaceSubsystem.java [hunk 7]
```diff
@@ -146,6 +155,7 @@ public TargetCubeInfo getTargetCube() {
     @Override
     public void updatePeriodicData() {
         sendWheelOdomUpdate();
+        sendScoringPlacement();
```
Comment (WasabiFan) [inline Competition2018/src/main/java/competition/subsystems/offboard/OffboardInterfaceSubsystem.java:158]: My understanding is that this data shouldn't change very often. Could we store the last placement data and only send when it changes? In the future we should probably make it possible to re-send it if the message is lost the first time, but for now that should work.

**Comment by JohnGilb** [comment]
Looks like this isn't used, closing for now.

**Pull request closed without merging**